tools/fs/files.js

```diff
@@ -723,6 +723,14 @@ files.extractTarGz = function (buffer, destPath, options) {
     // PaxHeader directory.
     .filter(file => ! file.startsWith("PaxHeader"));
 
+  // Sometimes windows fs does not catch up in time which makes readdir
+  // to return a wrong list of folders / files so we wait until
+  // the file system is in sync
+  if (topLevelOfArchive.length > 1 && process.platform === "win32") {
```

Comment:
Can you shed some light on what files are present in the directory? In other words, what are the additional files above the `1` (file) that we actually want?

Reply:
Thanks for all your comments @abernix !! I'll address them asap!

The files in question are described in the description of the pr, hopefully it makes sense! 

The thing is, at this point, we only expect the folder where we extracted the contents of the package to exist, but due to fs being out of sync, `readdir` returns the temporary files created in the extraction process, although they were successfully removed by `unlink`.

So let's say that:
```javascript
const tempDir = 'C:\Users\%userprofile%\AppData\Local\Temp\.tmp12zjjrp.xktr'
fs.readdir(tempDir); // Returns wrong list when fs is not in sync
```
So we expect `readdir` to return for example `['coffeescript-1.11.1_4']`, but instead it returns `['coffeescript-1.11.1_4', 'out.tar.gz', 'out.tar']` although the last two were already removed at that point. 
So the function that waits for fs to sync, performs `stat` operations over the elements of `readdir`'s output, which throw an exception that we catch when the file can't be accessed, in this case because it was already removed and wouldn't carry on with the execution until `readdir` returns a list of files / folders that are present on disk. In this case we expect `['coffeescript-1.11.1_4']`, but it could return any list of files, being the latter scenario caught later in https://github.com/meteor/meteor/blob/devel/tools/fs/files.js#L728 but now for good reasons.

Do you envisage any flaws in this process ?

Reply:
> The files in question are described in the description of the pr, hopefully it makes sense!

You mentioned:

> Files created in the 7z.exe extraction process of a package

...but it sounds like you're referring to just `out.tar.gz` and `out.tar`.  The `out.tar` file is the only file expected to be _created_ in that extraction process; the `tar.gz` is not created in the extraction, it's downloaded before the extraction begins.  Maybe I'm missing something?  If they're ending up in the same directory and that's causing the issue, perhaps there's another option where we extract into a separate directory from where we download?  Does that solve anything?

> Do you envisage any flaws in this process ?

Yes.  I understand how this is important and hope to get a solution here, but I think the other comments I left (on other lines of code) outline my concerns with a solution of the type you've offered here.  This is an aggressive file-system operation (running an unbounded number of additional file system read operations to wait for sync) and we should consider any other option before moving in this direction.  For example, the Node issues you linked to seem to suggest the option of using method which wait for sync from the file-system via [`fs.fsyncSync`](https://nodejs.org/dist/latest-v8.x/docs/api/fs.html#fs_fs_fsyncsync_fd)?  Have you looked at that?